src/REP15.sol

```diff
@@ -0,0 +1,466 @@
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function stopOwnershipDelegation(uint256 tokenId) external virtual {
+    REP15Utils.Delegation storage delegation = _delegations[tokenId];
+    address delegatee = delegation.delegatee;
+
+    if (!delegation.isActive()) revert REP15InactiveOwnershipDelegation(tokenId);
+
+    _checkAuthorizedDelegatee(delegatee, _msgSender(), tokenId);
+
+    delete _delegations[tokenId];
+
+    emit OwnershipDelegationStopped(tokenId, delegatee);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function createContext(address controller, uint64 detachingDuration, bytes calldata ctxMsg)
+    external
+    virtual
+    returns (bytes32 ctxHash)
+  {
+    ctxHash = keccak256(abi.encodePacked(_msgSender(), ctxMsg));
+
+    _updateContext(ctxHash, controller, detachingDuration, address(0));
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function updateContext(bytes32 ctxHash, address newController, uint64 newDetachingDuration) external virtual {
+    _updateContext(ctxHash, newController, newDetachingDuration, _msgSender());
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function attachContext(bytes32 ctxHash, uint256 tokenId, bytes calldata data) external virtual {
+    address operator = _msgSender();
+
+    _checkAuthorizedOwnershipManager(tokenId, operator);
+
+    _attachContext(ctxHash, tokenId, operator, data);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function requestDetachContext(bytes32 ctxHash, uint256 tokenId, bytes calldata data) external virtual {
+    address operator = _msgSender();
+
+    if (operator != _contexts[ctxHash].controller) {
+      _checkAuthorizedOwnershipManager(tokenId, operator);
+      _requestDetachContext(ctxHash, tokenId, operator, data);
+    } else {
+      _detachContext(ctxHash, tokenId, operator, data, false, true);
+    }
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function execDetachContext(bytes32 ctxHash, uint256 tokenId, bytes calldata data) external virtual {
+    address operator = _msgSender();
+
+    _checkAuthorizedOwnershipManager(tokenId, operator);
+
+    _detachContext(ctxHash, tokenId, operator, data, true, true);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function setContextLock(bytes32 ctxHash, uint256 tokenId, bool lock) external virtual {
+    _checkAuthorizedController(_msgSender(), ctxHash);
+
+    _requireAttachedTokenContext(ctxHash, tokenId, true).locked = lock;
+
+    emit ContextLockUpdated(ctxHash, tokenId, lock);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function setContextUser(bytes32 ctxHash, uint256 tokenId, address user) external virtual {
+    _checkAuthorizedController(_msgSender(), ctxHash);
+
+    _requireAttachedTokenContext(ctxHash, tokenId, false).user = user;
+
+    emit ContextUserAssigned(ctxHash, tokenId, user);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function maxDetachingDuration() public view virtual override returns (uint64) {
+    return _maxDetachingDuration;
```

Comment:
We can directly return the value of the maximum detaching duration. There's no need to store it as an immutable variable, right?

Reply:
Yes, but it should be a parameter in the constructor for easy testing, and flexibility in many use-case contexts.